driver/utils/unicode_converter.h

```diff
@@ -0,0 +1,540 @@
+        UnicodeConverter & src_converter, const std::basic_string_view<SourceCharType> & src,
+        std::basic_string<PivotCharType> & pivot,
+        UnicodeConverter & dest_converter, std::basic_string<DestinationCharType> & dest,
+        const bool ensure_src_signature,
+        const bool trim_dest_signature
+    );
+
+private:
+    UConverter * converter_ = nullptr;
+
+    std::string encoded_signature_to_prepend_;
+    std::vector<std::string> encoded_signatures_to_trim_;
+    std::size_t encoded_signatures_to_trim_max_size_ = 0;
+
+    std::string pivot_signature_to_prepend_;
+    std::vector<std::string> pivot_signatures_to_trim_;
+    std::size_t pivot_signatures_to_trim_max_size_ = 0;
+};
+
+inline const std::size_t UnicodeConverter::getEncodedMinCharSize() const {
+    return ucnv_getMinCharSize(converter_);
+}
+
+template <typename CharType>
+inline std::basic_string_view<CharType> UnicodeConverter::consumeEncodedSignature(const std::basic_string_view<CharType> & str) const {
+    if (!str.empty()) {
+        for (auto & signature : encoded_signatures_to_trim_) {
+            auto str_no_sig = ::consumeSignature(str, make_string_view(signature));
+            if (str_no_sig.size() < str.size())
+                return str_no_sig;
+        }
+    }
+
+    return str;
+}
+
+template <typename CharType>
+inline std::basic_string_view<CharType> UnicodeConverter::consumePivotSignature(const std::basic_string_view<CharType> & str) const {
+    if (!str.empty()) {
+        for (auto & signature : pivot_signatures_to_trim_) {
+            auto str_no_sig = ::consumeSignature(str, make_string_view(signature));
+            if (str_no_sig.size() < str.size())
+                return str_no_sig;
+        }
+    }
+
+    return str;
+}
+
+template <typename CharType>
+inline std::size_t UnicodeConverter::consumeEncodedSignatureInPlace(std::basic_string<CharType> & str) const {
+    if (!str.empty()) {
+        for (auto & signature : encoded_signatures_to_trim_) {
+            const auto symbols_consumed = ::consumeSignatureInPlace(str, make_string_view(signature));
+            if (symbols_consumed > 0)
+                return symbols_consumed;
+        }
+    }
+
+    return 0;
+}
+
+template <typename CharType>
+inline std::size_t UnicodeConverter::consumePivotSignatureInPlace(std::basic_string<CharType> & str) const {
+    if (!str.empty()) {
+        for (auto & signature : pivot_signatures_to_trim_) {
+            const auto symbols_consumed = ::consumeSignatureInPlace(str, make_string_view(signature));
+            if (symbols_consumed > 0)
+                return symbols_consumed;
+        }
+    }
+
+    return 0;
+}
+
+template <typename CharType, typename PivotCharType>
+inline void UnicodeConverter::convertToPivot(
+    const std::basic_string_view<CharType> & encoded,
+    std::basic_string<PivotCharType> & pivot,
+    const bool ensure_encoded_signature,
+    const bool trim_pivot_signature
+) {
+    pivot.clear();
```

Comment:
Please add a simplified plan/high-level overview of what is happening here in overly simplified pseudocode or just plain English. Hope it would make the reader's life easier.

Reply:
Done.